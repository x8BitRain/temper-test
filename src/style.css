@import 'css/animations.scss';
@import 'css/buttons.scss';
@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;
  background-color: #f6f6f6;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

body, html {
  height: 100vh;
  width: 100vw;
  max-width: 100%;
  padding: 0;
  margin: 0;
  overflow: hidden;
}

#app {
  height: 100%;
  width: 100%;
  max-width: 100%;
  margin: 0 auto;
  padding: 2rem;
  overflow: hidden;
}
