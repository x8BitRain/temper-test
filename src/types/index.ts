export enum MoveDirections {
  UP = 'up',
  DOWN = 'down',
}

export interface Post {
  id: string
  title: string
  content: string
}

export interface HistoryItem {
  description: string
  historySnapshot: Post[]
}

export type PostsStore = {
  posts: Post[]
  history: HistoryItem[]
}
