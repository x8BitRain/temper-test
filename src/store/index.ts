import { defineStore } from 'pinia'
import { MoveDirections, Post, PostsStore } from '../types'

const POSTS_URL = 'https://jsonplaceholder.typicode.com/posts?_limit=5'

export const usePostsStore = defineStore('posts', {
  state: () =>
    ({
      posts: [],
      history: [],
      error: null,
    } as PostsStore),
  actions: {
    async getPosts() {
      try {
        const request = await fetch(POSTS_URL)
        this.posts = await request.json()
      } catch (error: any | Error) {
        this.handleError(error)
      }
    },
    timeTravel(selectedIndex: number) {
      // Sets the posts state to a saved history snapshot by index
      // and removes history items before the selected one
      this.posts = this.history[selectedIndex].historySnapshot
      this.history = this.history.slice(selectedIndex + 1)
    },
    addHistoryItem(
      currentIndex: number,
      newIndex: number,
      post: Post,
      historySnapshot: Post[]
    ) {
      // Appends a history item object to the history state with string description
      const historyItem = {
        description: `Moved post ${post.id} from index ${currentIndex} to index ${newIndex}`,
        historySnapshot,
      }
      this.history?.unshift(historyItem)
    },
    movePost(direction: MoveDirections, currentIndex: number) {
      // Clones the current post state and moves the current post by index forward or backward
      // in the posts state depending on direction then constructs a history item with addHistoryItem()
      const currentHistory = [...this.posts]
      const currentPost = this.posts?.splice(currentIndex, 1)[0]
      const newIndex =
        direction === MoveDirections.UP ? currentIndex - 1 : currentIndex + 1
      this.posts?.splice(newIndex, 0, currentPost)
      this.addHistoryItem(currentIndex, newIndex, currentPost, currentHistory)
    },
    handleError(error: string) {
      this.error = error?.toString()
      throw new Error(<string>error)
    },
  },
})
