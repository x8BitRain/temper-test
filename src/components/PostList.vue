<script setup lang="ts">
  import Post from './Post.vue'
  import { usePostsStore } from '../store'
  import { storeToRefs } from 'pinia'
  const store = usePostsStore()
  const { posts } = storeToRefs(store)
</script>

<template>
  <div>
    <h1>Sortable Post List</h1>
    <div class="block max-w-md">
      <Post
        v-for="(post, index) in posts"
        :post="post"
        :index="index"
        :key="post.id"
      />
    </div>
  </div>
</template>

<style scoped></style>
