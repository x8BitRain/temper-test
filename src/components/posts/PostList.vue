<script setup lang="ts">
  import Post from './Post.vue'
  import { usePostsStore } from '../../store'
  import { storeToRefs } from 'pinia'

  const store = usePostsStore()
  const { posts } = storeToRefs(store)
</script>

<template>
  <div class="tt-post-list">
    <h1 class="tt-post-list__header">Sortable Post List</h1>
    <div>
      <transition-group
        name="posts-list-order"
        tag="div"
        v-if="posts.length"
      >
        <Post
          v-for="(post, index) in posts"
          :post="post"
          :index="index"
          :key="post.id"
        />
      </transition-group>
    </div>
  </div>
</template>

<style scoped lang="scss">
  .tt-post-list {
    @apply block flex-1;
    &__header {
      @apply text-3xl font-semibold py-4 text-white;
    }
  }
</style>
