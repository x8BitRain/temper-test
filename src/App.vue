<script setup lang="ts">
  import { onBeforeMount } from 'vue'
  import { usePostsStore } from './store'
  import {storeToRefs} from "pinia";
  import ActionList from './components/actions/ActionList.vue'
  import BackgroundSlice from './components/misc/BackgroundSlice.vue'
  import Error from "./components/Error.vue";
  import PostList from './components/posts/PostList.vue'

  const store = usePostsStore()
  const { error } = storeToRefs(store)

  onBeforeMount(async () => {
    if (import.meta.vitest) return // If testing don't get posts
    await store.getPosts()
  })
</script>

<template>
  <div v-if="!error" class="tt-app-container">
    <BackgroundSlice />
    <PostList />
    <ActionList />
  </div>
  <Error v-else :error="error"/>
</template>

<style scoped lang="scss">
  .tt-app-container {
    @apply flex justify-around max-w-5xl m-auto min-w-xsm gap-16 drop-shadow-md h-full max-md:block;
  }
</style>
