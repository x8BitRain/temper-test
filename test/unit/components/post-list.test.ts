import { mount } from '@vue/test-utils'
import { describe, expect, it, beforeEach } from 'vitest'
import PostList from '../../../src/components/posts/PostList.vue'
import { setActivePinia, createPinia } from 'pinia'
import { usePostsStore } from '../../../src/store'
import { postData } from '../../mock/data'
import { postObjects } from '../../helpers/pageObjects'

const DOWN_BUTTON_COUNT = 4
const UP_BUTTON_COUNT = 4

describe('PostList.vue', () => {
  beforeEach(async () => {
    setActivePinia(createPinia())
    const store = usePostsStore()
    store.posts = postData // Mock post data
  })
  it('should render the correct amount of up/down buttons', async () => {
    const postList = await mount(PostList)
    const visibleUpButtons = postList
      .findAll(postObjects.upButton)
      .map((btn) => btn.isVisible())
    const visibleDownButtons = postList
      .findAll(postObjects.downButton)
      .map((btn) => btn.isVisible())

    expect(visibleUpButtons.length).toEqual(UP_BUTTON_COUNT)
    expect(visibleDownButtons.length).toEqual(DOWN_BUTTON_COUNT)
  })

  it('should change the order of posts when clicking DOWN button', async () => {
    const postList = await mount(PostList)
    let firstPostElement = postList.findAll(postObjects.root)[0]
    const firstPostTitle = postData[0].title
    const secondPostTitle = postData[1].title
    const downButton = firstPostElement.find(postObjects.downButton)

    expect(firstPostElement.exists()).toBe(true)
    expect(firstPostElement.text()).toContain(firstPostTitle)
    expect(downButton.exists()).toBe(true)

    await downButton.trigger('click')
    // Get first post again after clicking down button
    firstPostElement = postList.findAll(postObjects.root)[0]
    expect(firstPostElement.text()).toContain(secondPostTitle)
  })

  it('should change the order of posts when clicking UP button', async () => {
    // First element should contain title of second element after clicking UP button
    const postList = await mount(PostList)
    let secondPostElement = postList.findAll(postObjects.root)[1]
    const secondPostTitle = postData[1].title
    const upButton = secondPostElement.find(postObjects.upButton)

    expect(secondPostElement.exists()).toBe(true)
    expect(secondPostElement.text()).toContain(secondPostTitle)
    expect(upButton.exists()).toBe(true)

    await upButton.trigger('click')
    // Get first post again after clicking up button
    const firstPostElement = postList.findAll(postObjects.root)[0]
    expect(firstPostElement.text()).toContain(secondPostTitle)
  })
})
