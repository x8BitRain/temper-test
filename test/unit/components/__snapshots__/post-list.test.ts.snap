// Vitest Snapshot v1

exports[`App.vue > should render 1`] = `
"<div class=\\"tt-app-container\\" data-v-7a7a37b1=\\"\\">
  <div class=\\"tt-background-slice\\" data-v-3ea0bed3=\\"\\" data-v-7a7a37b1=\\"\\"></div>
  <div class=\\"tt-post-list\\" data-v-60674a44=\\"\\" data-v-7a7a37b1=\\"\\">
    <h1 class=\\"tt-post-list__header\\" data-v-60674a44=\\"\\">Sortable Post List</h1>
    <div data-v-60674a44=\\"\\"><svg class=\\"tt-icon\\" viewBox=\\"0 0 24 24\\" width=\\"1.2em\\" height=\\"1.2em\\" data-v-f217e8d9=\\"\\" data-v-60674a44=\\"\\">
        <path fill=\\"currentColor\\" d=\\"M12 4V2A10 10 0 0 0 2 12h2a8 8 0 0 1 8-8Z\\"></path>
      </svg></div>
  </div>
  <div class=\\"tt-action-list\\" data-v-d0f28be6=\\"\\" data-v-7a7a37b1=\\"\\">
    <div class=\\"tt-action-list__header-container\\" data-v-d0f28be6=\\"\\">
      <h1 class=\\"tt-action-list__header-text\\" data-v-d0f28be6=\\"\\">List of actions committed</h1>
    </div>
    <div class=\\"tt-action-list__item-container\\" data-v-d0f28be6=\\"\\">
      <div class=\\"tt-action-item-placeholder\\" data-v-cea37cc8=\\"\\" data-v-d0f28be6=\\"\\">
        <div data-v-cea37cc8=\\"\\">No actions yet...</div>
      </div>
    </div>
  </div>
</div>"
`;
